dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs

```diff
@@ -94,100 +95,277 @@ public Plan CreatePlan(string goal)
         return plan;
     }
 
+    /// <summary>
+    /// Execute a plan
+    /// </summary>
+    /// <param name="question">The question to answer</param>
+    /// <param name="context">The context to use</param>
+    /// <param name="token">The cancellation token</param>
+    /// <returns>The context with the result</returns>
+    /// <exception cref="SKException">No AIService available for getting completions.</exception>
     [SKFunction, SKName("ExecutePlan"), Description("Execute a plan")]
     public async Task<SKContext> ExecutePlanAsync(
         [Description("The question to answer")]
         string question,
-        [Description("List of tool descriptions")]
-        string functionDescriptions,
-        SKContext context)
+        SKContext context,
+        CancellationToken token = default)
     {
+        if (string.IsNullOrEmpty(question))
+        {
+            context.Variables.Update("Question not found.");
+            return context;
+        }
+
         var stepsTaken = new List<SystemStep>();
-        if (!string.IsNullOrEmpty(question))
+
+        ChatHistory chatHistory = await this.InitializeChatHistoryAsync(this.CreateChatHistory(this._kernel, out var aiService), aiService, context).ConfigureAwait(false);
+        var chatCompletion = aiService as IChatCompletion;
+        var textCompletion = aiService as ITextCompletion;
+
+        var startingMessageCount = chatHistory.Messages.Count;
+
+        var GetMessageTokens = () =>
         {
-            for (int i = 0; i < this.Config.MaxIterations; i++)
+            var messages = string.Join("\n", chatHistory?.Messages);
+            var tokenCount = messages.Length / 4;
+            return tokenCount;
+        };
+
+        var addThought = true;
+        var GetCompletionAsync = async () =>
+        {
+            if (chatCompletion is not null)
```

Comment:
No need to declare `chatCompletion` above and check if it has IChatCompletion type. It can easily be done in one line - `if (aiService is IChatCompletion chatCompletion)`